src/topics/async.md

```diff
@@ -0,0 +1,461 @@
+keep running, waiting for page requests, even though the main program has
+reached the end.
+
+The "event loop" has that name because it is a "loop" which runs continuously,
+looking for "events" which have happened and processing them. It will continue
+to loop until it is no longer expecting more events to happen.
+
+The event loop works something like the following pseudo-code:
+
+```js
+while (true) {
+  for (each event that happened) {
+    process the event
+  }
+  if (no more pending events) {
+    break;
+  }
+}
+```
+
+Some examples of events for which the event loop will wait:
+
+* An asynchronous filesystem operation has completed.
+* An asynchronous network operation has completed.
+* A new incoming network connection was detected.
+* A timer has timed out.
+
+There are two functions which allow you to directly manipulate the event loop:
+`nextTick()` and `setImmediate()`.
+
+The `nextTick()` function allows you to give a callback which is added to the
+beginning of the list of completed events. Said another way, the callback you
+pass to `nextTick()` will be called as soon as the current event handler is
+done.
+
+The `setImmediate()` function allows you to give a callback which is processed
+the next time through the loop.
+
+The distinction between `nextTick()`, `setImmediate()`, and
+`setTimeout(..., 0)` is subtle.
+
+
+### Exceptions and Stack Traces
+Since an asynchronous function can call the callback some time after it was
+invoked, the execution environment of the callback can be quite different than
+what existed when the async function was called.
+
+Consider the following example:
+
+```js
+function err() {
+    var e = new Error('boom');
+    console.log(e.stack);
+}
+function run() {
+    err();
+    setTimeout(err, 100);
+}
+run();
+```
+
+This has the following output:
+
+```
+Error: boom
+    at err (/Users/drewfish/stacktrace.js:2:13)
+    at run (/Users/drewfish/stacktrace.js:6:5)
+    at Object.<anonymous> (/Users/drewfish/stacktrace.js:9:1)
+    at Module._compile (module.js:456:26)
+    at Object.Module._extensions..js (module.js:474:10)
+    at Module.load (module.js:356:32)
+    at Function.Module._load (module.js:312:12)
+    at Function.Module.runMain (module.js:497:10)
+    at startup (node.js:119:16)
+    at node.js:929:3
+Error: boom
+    at err [as _onTimeout] (/Users/drewfish/stacktrace.js:2:13)
+    at Timer.listOnTimeout [as ontimeout] (timers.js:112:15)
+```
+
+Note how the stacktrace from the first call to `err()` does have the `run()`
+function in it, yet the second call doesn't. This is because the second call
+to `err()` happens _after_ `run()` has completed so `run()` is no longer part
+of the execution environment.
+
+This is where the event loop shows up: the second call to `err()` happens
+during a timeout which is handled in the event loop.
+
+
+#### Throwing and try/catch
+One thing to be careful about in asynchronous programming is throwing errors
```

Comment:
Promises are throw safe - it is important to mention with promises you can try/catch in asynchronous code.